src/gwt/src/org/rstudio/studio/client/workbench/views/output/lint/LintManager.java

```diff
@@ -0,0 +1,315 @@
+import org.rstudio.studio.client.common.RetinaStyleInjector;
+import org.rstudio.studio.client.common.filetypes.TextFileType;
+import org.rstudio.studio.client.server.ServerError;
+import org.rstudio.studio.client.server.ServerRequestCallback;
+import org.rstudio.studio.client.server.Void;
+import org.rstudio.studio.client.workbench.prefs.model.UIPrefs;
+import org.rstudio.studio.client.workbench.views.output.lint.model.LintItem;
+import org.rstudio.studio.client.workbench.views.output.lint.model.LintServerOperations;
+import org.rstudio.studio.client.workbench.views.source.editors.text.DocDisplay;
+import org.rstudio.studio.client.workbench.views.source.editors.text.TextEditingTarget;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.Position;
+import org.rstudio.studio.client.workbench.views.source.editors.text.cpp.CppCompletionRequest;
+import org.rstudio.studio.client.workbench.views.source.events.SourceFileSavedEvent;
+import org.rstudio.studio.client.workbench.views.source.events.SourceFileSavedHandler;
+import org.rstudio.studio.client.workbench.views.source.model.CppDiagnostic;
+
+import com.google.gwt.core.client.JsArray;
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
+import com.google.gwt.user.client.Command;
+import com.google.gwt.user.client.Timer;
+import com.google.inject.Inject;
+
+public class LintManager
+{
+   class LintContext
+   {
+      public LintContext(Invalidation.Token token,
+                         Position cursorPosition,
+                         boolean showMarkers,
+                         boolean excludeCurrentStatement)
+      {
+         this.cursorPosition = cursorPosition;
+         this.token = token;
+         this.showMarkers = showMarkers;
+         this.excludeCurrentStatement = excludeCurrentStatement;
+      }
+      
+      public final Invalidation.Token token;
+      public final Position cursorPosition;
+      public final boolean showMarkers;
+      public final boolean excludeCurrentStatement;
+   }
+   
+   private void reset()
+   {
+      showMarkers_ = false;
+      excludeCurrentStatement_ = true;
+   }
+   
+   // NOTE: by 'lintable' we mean 'uses RStudio-internal' linter
+   // rather than Ace worker
+   private boolean isLintableDocument()
+   {
+      TextFileType type = docDisplay_.getFileType();
+      return type.isR() ||
+             type.isC() ||
+             type.isCpp() ||
+             type.isRmd() ||
+             type.isRnw() ||
+             type.isRpres();
+   }
+   
+   public LintManager(TextEditingTarget target)
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      target_ = target;
+      docDisplay_ = target.getDocDisplay();
+      showMarkers_ = false;
+      invalidation_ = new Invalidation();
+      timer_ = new Timer()
+      {
+         
+         @Override
+         public void run()
+         {
+            if (!isLintableDocument())
```

Comment:
If the user switches document types (using the gutter toolbar) do we need to clear any existing lint?